### PR TITLE
Document workings of integration tests

### PR DESCRIPTION
**Pull request by Motiejus**
Did some best-effort scribbling after clarifying the test structure with @CorgiMan .

### CODE REVIEW

**Comment by dansimau** [comment]
Love the documentation :+1:

File: test/README.md [hunk 1]
```diff
@@ -65,3 +65,34 @@ Install ringpop-go and build the `testpop` executable:
...
+
+# Understanding the integration tests
+
+This is a written summary of a one-hour workshop of digging into the integration tests and asking questions from the people who implemented it. As of now, it might be out of date (we will try to keep it up to date though), but the text below can give a good starting point while trying to wrap your head around.
+
+## Glossary
+
+*Slanted names* mean general concepts, names in `monospace` common variable names.
+
+* *SUT*: Subject Under Test.
```
Comment (CorgiMan) [inline test/README.md]: I think we should add: Also referred to as "the real node"

File: test/README.md [hunk 2]
```diff
@@ -65,3 +65,34 @@ Install ringpop-go and build the `testpop` executable:
...
+* *testpop*: very basic application using ringpop.
```
Comment (CorgiMan) [inline test/README.md]: I think we should add: testpop is commonly used by tickcluster and acts as the SUT for the integration test

File: test/README.md [hunk 3]
```diff
@@ -65,3 +65,34 @@ Install ringpop-go and build the `testpop` executable:
...
+* `n` : size of the cluster.
+* `ns`: list of cluster sizes. Treat it like *plural n* -- *ns*.
+* `nodeIx`: node index. Any variable that ends with `Ix` is an index variable.
+* `cb`: callback.
+
+## High-level overview
+
+Tests are usually composed of a *Real Node* (*SUT*) and a number of fake nodes. The real node is the actual ringpop instance (`testpop`), and the fake nodes are the test harness.
+
+The test harness (fake nodes) send messages to the real node, and listen on SUTS sends back. Test harness then asserts whether whatever SUT is sending back matches expectations. With this black-box structure, tests are verifying the node behaves as expected given certain messages, and can be used to achieve feature parity between different ringpop implementations.
```
Comment (CorgiMan) [inline test/README.md]: on 'what' the SUT sends back

File: test/ringpop-assert.js [hunk 4]
```diff
@@ -669,9 +669,26 @@ function createValidateEvent(t, tc) {
     };
 }
 
-// validates a scheme on incoming events send by the real-node. A scheme is a collection of
-// functions from scheme.js. On every incoming event we try to progress through the scheme.
-// further. When all the functions in the scheme have ran, the test is a success.
+// validates a scheme on incoming events send by the real-node. A scheme is a
+// collection of functions from scheme.js. On every incoming event we try to
```
Comment (CorgiMan) [inline test/ringpop-assert.js]: s/scheme.js/ringpop-assert.js/

File: test/ringpop-assert.js [hunk 5]
```diff
@@ -669,9 +669,26 @@ function createValidateEvent(t, tc) {
...
+// progress through the scheme further. When all the functions in the scheme
+// have ran, the test is a success.
+//
+// A scheme is the list of functions passed by the caller. Every function in
+// the list will be called like this:
+// - fun(eventlist, callback)
+//
+// fun needs to verify if eventlist is what it is expecting to be. If it is,
+// fun calls `cb(list)`, where list is a filtered list. Filtered list is what
```
Comment (CorgiMan) [inline test/ringpop-assert.js]: The discarded items are no longer necessary in the tests.

File: test/ringpop-assert.js [hunk 6]
```diff
@@ -669,9 +669,26 @@ function createValidateEvent(t, tc) {
...
+// the function processed, and is no longer necessary in the tests. For
+// example, a closure which checks for 3 ping requests finds three ping
+// requests, removes them from the list, and calls the callback. If the
+// callback is called with a list, the test will go on to the next closure.
+//
+// If the list is not complete, call `cb(null)`; then the fun will be called
+// again after some time, with more elements in the list.
```
Comment (CorgiMan) [inline test/ringpop-assert.js]: the fun will be called again when more elements are added to the list.

File: test/README.md (hunk 1 above)
Comment (Motiejus) [inline test/README.md]: :+1:

File: test/README.md (hunk 2 above)
Comment (Motiejus) [inline test/README.md]: :+1:

File: test/ringpop-assert.js (hunk 4 above)
Comment (Motiejus) [inline test/ringpop-assert.js]: :+1:

File: test/ringpop-assert.js (hunk 5 above)
Comment (Motiejus) [inline test/ringpop-assert.js]: Reworded.

File: test/ringpop-assert.js (hunk 6 above)
Comment (Motiejus) [inline test/ringpop-assert.js]: :+1:

**Comment by Motiejus** [comment]
@CorgiMan : addressed all your feedback in the last commit.

**Comment by CorgiMan** [comment]
LGTM

**Comment by Motiejus** [comment]
Merging the squashed version ...

**Pull request merged**